app/assets/javascripts/event-visualization.js

```diff
@@ -1,92 +1,118 @@
-function createBlankNetworkDiagram(elementId) {
-  const nodes = new vis.DataSet([]);
-  const edges = new vis.DataSet([]);
-  var container = document.getElementById(elementId);
-  var data = {
-    nodes: nodes,
-    edges: edges
+var EventDiagram = function()
+{
+
+  var addBehaviorToTheDiagram = function(diagram)
+  {
+    var buildStepNodeObject = function(step, properties) {
+      const { name, type, guid, next_steps } = step;
+      return Object.assign({
+        id: diagram.nodes.length,
+        label: `${name}:${type}:${guid}`,
+        widthConstraint: 60,
+        shape: 'box',
+        step: step
+      }, properties);
+    };
+
+    diagram.addEvent = function (event) {
+      diagram.edges.add({ from: event.id, to: event.prior_event_id });
+
+      diagram.nodes.add({
+        id: event.id,
+        label: event.message,
+        priorEventId: event.prior_event_id
+      });
+    };
+
+    diagram.addStepAsNode = function(step, parentNode) {
+      const { name, type, guid, next_steps } = step;
+      const newNode = buildStepNodeObject(step, { id: diagram.nodes.length });
+
+      if (parentNode) {
+        const newEdge = { from: parentNode.id, to: newNode.id, arrows: { to: true } };
+        diagram.edges.add(newEdge);
+        newNode.parentNode = parentNode;
+      }
+
+      diagram.nodes.add(newNode);
+
+      if (step.next_steps) {
+        const nextSteps = step.next_steps.forEach(step => diagram.addStepAsNode(step, newNode));
+        return [newNode].concat(nextSteps);
+      } else {
+        return [newNode];
+      }
+
+      return newNode;
+    };
+
+    diagram.addEvents = function(events) {
+      events.forEach(x => { diagram.addEvent(x); });
+    };
+
+    return diagram;
   };
-  var options = {};
-  var network = new vis.Network(container, data, options);
-
-  return { network, nodes, edges }
-}
-
-function showEventBubbles(diagram, events) {
-  events.forEach(thisEvent => {
-    addEventInstanceAsIndividualNode(diagram, thisEvent);
-  });
-}
-
-function addEventInstanceAsIndividualNode(diagram, event) {
-  diagram.edges.add({ from: event.id, to: event.prior_event_id });
-
-  diagram.nodes.add({
-    id: event.id,
-    label: event.message,
-    priorEventId: event.prior_event_id
-  });
-}
-
-var addEventInstanceAsAggregateNodes = function() {
```

Comment:
Some stuff went completely away. like this.

Reply:
It didn't go away -- it was just moved here:  https://github.com/darrencauthon/fasten/pull/53/files#diff-cd4ecb39b98c53b31ff9e2a12af30838R81

Much of the code between showing raw events and aggregate events was exactly the same.  The only difference I saw is that the raw events just converted the event to a node, while the aggregate added logic to aggregate the events by type.

So when I refactored this code, I made an EventDiagram to show events, and an AggregateEventDiagram to show aggregates.  Look at the index.html and aggregate.html and notice that they look exactly the same, except for that one class difference.